docs/design/proposals/resource-reference/resource-reference.md

```diff
@@ -0,0 +1,555 @@
+		FieldConfig:       nil,
+	}
+}
+```
+
+#### Runtime Changes
+
+```go
+// aws_resource_manager.go
+
+type AWSResourceManager interface {
+        ...
++       // ResolveReferences finds if there are any AWSResourceReference fields
++       // present in the AWSResource passed in the parameters and attempts to resolve
++       // those reference fields into target field.
++       // It returns an AWSResource with resolved references, boolean representing
++       // whether any reference fields were present and an error if the passed
++       // AWSResource's reference fields cannot be resolved
++       ResolveReferences(context.Context, client.Reader, AWSResource) (AWSResource, bool, error)
+}
+```
+
+```go
+// reconciler.go
+
+// reconcile either cleans up a deleted resource or ensures that the supplied
+// AWSResource's backing API resource matches the supplied desired state.
+//
+// It returns a copy of the resource that represents the latest observed state.
+func (r *resourceReconciler) reconcile(
+ctx context.Context,
+rm acktypes.AWSResourceManager,
+res acktypes.AWSResource,
+) (acktypes.AWSResource, error) {
+    if res.IsBeingDeleted() {
++       // Resolve references before deleting the resource.
++       // Ignore any errors while resolving the references
++    	  res, _, _ = rm.ResolveReferences(ctx, r.apiReader, res)
+        return r.deleteResource(ctx, rm, res)
+    }
+    return r.Sync(ctx, rm, res)
+}
+
+// Sync ensures that the supplied AWSResource's backing API resource
+// matches the supplied desired state.
+//
+// It returns a copy of the resource that represents the latest observed state.
+func (r *resourceReconciler) Sync(
+	ctx context.Context,
+	rm acktypes.AWSResourceManager,
+	desired acktypes.AWSResource,
+) (acktypes.AWSResource, error) {
+	var err error
+	rlog := ackrtlog.FromContext(ctx)
+	exit := rlog.Trace("r.Sync")
+	defer exit(err)
+
+	var latest acktypes.AWSResource // the newly created or mutated resource
+
+	isAdopted := IsAdopted(desired)
+	rlog.WithValues("is_adopted", isAdopted)
++	rlog.Enter("rm.ResolveReferences")
++	resolvedRefDesired, containsResourceRef, err := rm.ResolveReferences(ctx, r.apiReader, desired)
++	rlog.Exit("rm.ResolveReferences", err)
++	if err != nil {
++		// Set the condition in copy of desired resources so that it can be
++		//patched back in etcd
++		desiredCopy := desired.DeepCopy()
++		if containsResourceRef {
++			condition.SetReferenceResolved(desiredCopy, corev1.ConditionFalse,
++				&condition.ReferenceNotResolvedMessage,
++				&condition.ReferenceNotResolvedReason)
++		}
++		return desiredCopy, err
++	}
+
++	// If the desired resource had references, set the ReferenceResolved
++	//condition to true
++	if containsResourceRef {
++		condition.SetReferenceResolved(resolvedRefDesired, corev1.ConditionTrue,
++			&condition.ReferenceResolvedMessage,
++			&condition.ReferenceResolvedReason)
++	} else {
++		condition.RemoveReferenceResolved(resolvedRefDesired)
++	}
```

Comment:
do all of this condition-setting within the `AWSResourceResolver.ResolveReferences` method implementation...